### ISSUE TITLE
Driver.close() behavior for main window and child window differs versus legacy firefox driver(v47)

### ISSUE THREAD

**Issue by soumyababy81** [mozilla/geckodriver#1146]
## System
* Version: 19/19.1
* Platform: Windows 10
* Firefox: Firefox v58 and Nightly build v59.0a1
* Selenium: 3.8.1

Scenario 1.
I have a page(main window) in which some changes are done and before submitting the changes for the page, I try to close the window using driver.close(). The browser is closed ignoring the alert (Leave page/Stay on Page). But if I try to use driver.quit, the warning/alert message pops up and its hanged and does not move to the next test. 

Workaround - use driver.close() ignoring the alert (as the test is not expecting to handle the alert) to move on to next test.

Legacy firefox(FFv47- driver.close() doesn’t ignore the alert and doesn’t hangs the test also when using driver.quit it exit with proper exception - Exception in thread "main" org.openqa.selenium.remote.SessionNotFoundException: The FirefoxDriver cannot be used after quit() was called.) 

Scenario 2.
I have a page which opens the second window and some changes are made on the second page. While trying to close the window using driver.close() the warning/alert message pops up(expected) but where as in above scenario driver.close() ignored the warning and closed. This is application’s expected behavior (warning/alert pops up) and it hangs and doesn't move to code to handle the alert pop up. 

Workaround – had to use robot class to close the second window with alert pop up and handle the alert message.

Legacy firefoxv47 - driver.close() pops up the alert and test is able to handle the alert. 

Both the above scenarios works perfect in legacy firefox driver(FF v47).

Example for above test scenario 1

System.setProperty("webdriver.gecko.driver","C:\\Git\\geckodriver.exe");
WebDriver driver = new FirefoxDriver();
driver.get("http://www.4guysfromrolla.com/demos/OnBeforeUnloadDemo1.htm");
driver.findElement(By.xpath("//input[@name='name']")).sendKeys("Test");
//driver.close();//Closes the browser ignoring the alert, where in legacy FF – doesn’t ignore the alert and below alert code steps are executed.
	
Alert alert=driver.switchTo().alert();
System.out.println(alert.getText());
alert.accept();

Please let us know if we are missing something here or what is the optimal way to handle above scenarios.

**Comment by andreastt** [comment]
The global modal dialogue “Are you sure you want to leave the page?” that appears if you try to close a document where you have modified a form control element is not supposed to be taken into consideration according to the [WebDriver standard](https://w3c.github.io/webdriver/webdriver-spec.html).

Note that WebDriver doesn’t have a `quit()` API like Selenium does,  but I’m assuming that is mapped down to something like calling Close Window and then Delete Session?

The Close Window command will delete the session implicitly and shut down the browser if there is only one window open. Subsequent calls to Close Window after the session has deleted will result in an invalid session ID error, which is the same for Delete Session.

To summarise, you should not expect to have to handle the global modal dialogue when closing a window that has unsubmitted form control data.

Does this address your concerns?

**Comment by andreastt** [comment]
The session is only deleted if there is _one_ window left open. It isn’t if you have two and one is closed.

**Comment by soumyababy81** [comment]
Yes I have two windows opened, I need to close the second window which pops up the alert message and I need to handle alert and then come back to main window. The driver.close() tries to close the second window (session is not deleted) popping up the alert message and its hanging there without executing the subsequent calls. How to handle this?

**Comment by soumyababy81** [comment]
Any thoughts on above points. Above scenario is taken care and working in legacy firefoxdriver(vFF47), so there should be a similar way for the latest gecko driver to handle. Let us know how to handle this and there should be a way to handle this.

**Comment by whimboo** [comment]
@soumyababy81 can you please provide a trace level log we could have a look at? Instructions can be found here: https://firefox-source-docs.mozilla.org/testing/geckodriver/geckodriver/TraceLogs.html.

It's actually an interesting behavior.

**Comment by soumyababy81** [comment]
@whimboo I have attached two simple html pages which can be used for testing. As well below is the sample code for testing the attached html pages.

Sample code:
System.setProperty("webdriver.gecko.driver","C:\\Git\\geckodriver.exe");
	     
		FirefoxOptions options = new FirefoxOptions();
		options.setLogLevel(FirefoxDriverLogLevel.TRACE);
		WebDriver driver = new FirefoxDriver(options);
	  
	  	driver.get("file:///C:/Users/soumyab/Desktop/Test/openwindow.htm");
		Thread.sleep(2000);
		
		driver.findElement(By.xpath("//button[text()='Open Window']")).click();
		  
		Set<String> allWindowHandles = driver.getWindowHandles();

		for (String handle : allWindowHandles) {
				driver.switchTo().window(handle);
		}
		
		driver.findElement(By.xpath("//input[@name='name']")).sendKeys("Hello");
		driver.close();
		 
		Alert alert=driver.switchTo().alert();
		System.out.println(alert.getText());
		alert.accept();

**TraceLog below**

1517313733618	geckodriver	INFO	geckodriver 0.19.1
1517313733625	geckodriver	INFO	Listening on 127.0.0.1:28674
1517313734214	mozrunner::runner	INFO	Running command: "C:\\Program Files (x86)\\Firefox Nightly\\firefox.exe" "-marionette" "-profile" "C:\\Users\\soumyab\\AppData\\Local\\Temp\\rust_mozprofile.nZSEQMNKRhhK"
1517313734722	Marionette	DEBUG	Received observer notification "profile-after-change"
1517313734870	Marionette	DEBUG	Received observer notification "command-line-startup"
1517313734870	Marionette	INFO	Enabled via --marionette
[DEBUG SHUTDOWN] Register: decoder=00000239AC5AE800, count=1
[DEBUG SHUTDOWN] ShutdownDecoder: decoder: 'wmf software video decoder - yuv420 (remote)' (00000239A69DACA0) flush:1
[DEBUG SHUTDOWN] operator (): pool=00000239A69D5DC0 shutdown=false count=0
1517313736228	geckodriver::marionette	TRACE	  connection attempt 0/600
1517313736285	Marionette	DEBUG	Received observer notification "sessionstore-windows-restored"
1517313736534	Marionette	DEBUG	Setting recommended pref toolkit.cosmeticAnimations.enabled to false
1517313736535	Marionette	DEBUG	Setting recommended pref datareporting.policy.dataSubmissionPolicyAccepted to false
1517313736536	Marionette	DEBUG	New connections are accepted
1517313736536	Marionette	INFO	Listening on port 58323
1517313737332	geckodriver::marionette	DEBUG	Connected to Marionette on localhost:58323
1517313737335	Marionette	DEBUG	Accepted connection 0 from 127.0.0.1:58347
1517313737336	geckodriver::marionette	TRACE	<- {"applicationType":"gecko","marionetteProtocol":3}
1517313737336	geckodriver::marionette	TRACE	-> 163:[0,1,"newSession",{"acceptInsecureCerts":true,"browserName":"firefox","capabilities":{"desiredCapabilities":{"acceptInsecureCerts":true,"browserName":"firefox"}}}]
1517313737338	Marionette	TRACE	0 -> [0,1,"newSession",{"acceptInsecureCerts":true,"browserName":"firefox","capabilities":{"desiredCapabilities":{"acceptInsecureCerts":true,"browserName":"firefox"}}}]
1517313737339	Marionette	WARN	TLS certificate errors will be ignored for this session
1517313737403	Marionette	DEBUG	Register listener.js for window 4294967297
1517313737408	Marionette	TRACE	0 <- [1,1,null,{"sessionId":"d96b0092-df51-4111-913c-771c8623e6d6","capabilities":{"browserName":"firefox","browserVersion":"59.0a ... sID":5900,"moz:profile":"C:\\Users\\soumyab\\AppData\\Local\\Temp\\rust_mozprofile.nZSEQMNKRhhK","moz:webdriverClick":true}}]
1517313737410	geckodriver::marionette	TRACE	<- [1,1,null,{"sessionId":"d96b0092-df51-4111-913c-771c8623e6d6","capabilities":{"browserName":"firefox","browserVersion":"59.0a1","platformName":"windows_nt","platformVersion":"10.0","pageLoadStrategy":"normal","acceptInsecureCerts":true,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"rotatable":false,"moz:accessibilityChecks":false,"moz:headless":false,"moz:processID":5900,"moz:profile":"C:\\Users\\soumyab\\AppData\\Local\\Temp\\rust_mozprofile.nZSEQMNKRhhK","moz:webdriverClick":true}}]
1517313737411	webdriver::server	DEBUG	<- 200 OK {"value": {"sessionId":"d96b0092-df51-4111-913c-771c8623e6d6","capabilities":{"acceptInsecureCerts":true,"browserName":"firefox","browserVersion":"59.0a1","moz:accessibilityChecks":false,"moz:headless":false,"moz:processID":5900,"moz:profile":"C:\\Users\\soumyab\\AppData\\Local\\Temp\\rust_mozprofile.nZSEQMNKRhhK","moz:webdriverClick":true,"pageLoadStrategy":"normal","platformName":"windows_nt","platformVersion":"10.0","rotatable":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000}}}}
Jan 30, 2018 5:32:17 PM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C
1517313737483	webdriver::server	DEBUG	-> POST /session/d96b0092-df51-4111-913c-771c8623e6d6/url {"url":"file:///C:/Users/soumyab/Desktop/Test/openwindow.htm"}
1517313737484	geckodriver::marionette	TRACE	-> 74:[0,2,"get",{"url":"file:///C:/Users/soumyab/Desktop/Test/openwindow.htm"}]
1517313737488	Marionette	TRACE	0 -> [0,2,"get",{"url":"file:///C:/Users/soumyab/Desktop/Test/openwindow.htm"}]
1517313737503	Marionette	DEBUG	Received DOM event "beforeunload" for "about:blank"
1517313737512	Marionette	DEBUG	Received DOM event "pagehide" for "about:blank"
1517313737513	Marionette	DEBUG	Received DOM event "unload" for "about:blank"
1517313737513	Marionette	DEBUG	Received observer notification "outer-window-destroyed" for "4294967297"
Unable to read VR Path Registry from C:\Users\soumyab\AppData\Local\openvr\openvrpaths.vrpath
[Parent 5900, Gecko_IOThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
Unable to read VR Path Registry from C:\Users\soumyab\AppData\Local\openvr\openvrpaths.vrpath
[Child 5624, Chrome_ChildThread] WARNING: pipe error: 109: file z:/build/build/src/ipc/chromium/src/chrome/common/ipc_channel_win.cc, line 346
1517313738049	Marionette	DEBUG	Register listener.js for window 8589934593
1517313738094	Marionette	DEBUG	Check readyState "uninitialized for "about:blank"
1517313738107	Marionette	DEBUG	Received DOM event "DOMContentLoaded" for "file:///C:/Users/soumyab/Desktop/Test/openwindow.htm"
1517313738160	Marionette	DEBUG	Received DOM event "pageshow" for "file:///C:/Users/soumyab/Desktop/Test/openwindow.htm"
1517313738161	Marionette	TRACE	0 <- [1,2,null,{}]
1517313738163	geckodriver::marionette	TRACE	<- [1,2,null,{}]
1517313738163	webdriver::server	DEBUG	<- 200 OK {"value": {}}
1517313740180	webdriver::server	DEBUG	-> POST /session/d96b0092-df51-4111-913c-771c8623e6d6/element {"value":"//button[text()='Open Window']","using":"xpath"}
1517313740181	geckodriver::marionette	TRACE	-> 78:[0,3,"findElement",{"using":"xpath","value":"//button[text()='Open Window']"}]
1517313740184	Marionette	TRACE	0 -> [0,3,"findElement",{"using":"xpath","value":"//button[text()='Open Window']"}]
1517313740191	Marionette	TRACE	0 <- [1,3,null,{"value":{"element-6066-11e4-a52e-4f735466cecf":"a025aefc-8477-4945-8308-2c8f3d037a47","ELEMENT":"a025aefc-8477-4945-8308-2c8f3d037a47"}}]
1517313740193	geckodriver::marionette	TRACE	<- [1,3,null,{"value":{"element-6066-11e4-a52e-4f735466cecf":"a025aefc-8477-4945-8308-2c8f3d037a47","ELEMENT":"a025aefc-8477-4945-8308-2c8f3d037a47"}}]
1517313740193	webdriver::server	DEBUG	<- 200 OK {"value":{"element-6066-11e4-a52e-4f735466cecf":"a025aefc-8477-4945-8308-2c8f3d037a47"}}
1517313740209	webdriver::server	DEBUG	-> POST /session/d96b0092-df51-4111-913c-771c8623e6d6/element/a025aefc-8477-4945-8308-2c8f3d037a47/click {"id":"a025aefc-8477-4945-8308-2c8f3d037a47"}
1517313740210	geckodriver::marionette	TRACE	-> 66:[0,4,"clickElement",{"id":"a025aefc-8477-4945-8308-2c8f3d037a47"}]
1517313740212	Marionette	TRACE	0 -> [0,4,"clickElement",{"id":"a025aefc-8477-4945-8308-2c8f3d037a47"}]
1517313740710	Marionette	DEBUG	Canceled page load listener because no navigation has been detected
1517313740711	Marionette	TRACE	0 <- [1,4,null,{}]
1517313740712	geckodriver::marionette	TRACE	<- [1,4,null,{}]
1517313740712	webdriver::server	DEBUG	<- 200 OK {"value": {}}
1517313740718	webdriver::server	DEBUG	-> GET /session/d96b0092-df51-4111-913c-771c8623e6d6/window/handles 
1517313740719	geckodriver::marionette	TRACE	-> 27:[0,5,"getWindowHandles",{}]
1517313740721	Marionette	TRACE	0 -> [0,5,"getWindowHandles",{}]
1517313740722	Marionette	TRACE	0 <- [1,5,null,["8589934593","8589934596"]]
1517313740722	geckodriver::marionette	TRACE	<- [1,5,null,["8589934593","8589934596"]]
1517313740723	webdriver::server	DEBUG	<- 200 OK {"value":["8589934593","8589934596"]}
1517313740731	webdriver::server	DEBUG	-> POST /session/d96b0092-df51-4111-913c-771c8623e6d6/window {"handle":"8589934593"}
1517313740732	geckodriver::marionette	TRACE	-> 44:[0,6,"switchToWindow",{"name":"8589934593"}]
1517313740734	Marionette	TRACE	0 -> [0,6,"switchToWindow",{"name":"8589934593"}]
1517313740735	Marionette	TRACE	0 <- [1,6,null,{}]
1517313740738	geckodriver::marionette	TRACE	<- [1,6,null,{}]
1517313740738	webdriver::server	DEBUG	<- 200 OK {"value": {}}
1517313740745	webdriver::server	DEBUG	-> POST /session/d96b0092-df51-4111-913c-771c8623e6d6/window {"handle":"8589934596"}
1517313740745	geckodriver::marionette	TRACE	-> 44:[0,7,"switchToWindow",{"name":"8589934596"}]
1517313740747	Marionette	TRACE	0 -> [0,7,"switchToWindow",{"name":"8589934596"}]
1517313740749	Marionette	DEBUG	Register listener.js for window 8589934596
1517313740751	Marionette	TRACE	0 <- [1,7,null,{}]
1517313740751	geckodriver::marionette	TRACE	<- [1,7,null,{}]
1517313740751	webdriver::server	DEBUG	<- 200 OK {"value": {}}
1517313740756	webdriver::server	DEBUG	-> POST /session/d96b0092-df51-4111-913c-771c8623e6d6/element {"value":"//input[@name='name']","using":"xpath"}
1517313740757	geckodriver::marionette	TRACE	-> 69:[0,8,"findElement",{"using":"xpath","value":"//input[@name='name']"}]
1517313740758	Marionette	TRACE	0 -> [0,8,"findElement",{"using":"xpath","value":"//input[@name='name']"}]
1517313740761	Marionette	TRACE	0 <- [1,8,null,{"value":{"element-6066-11e4-a52e-4f735466cecf":"dcdf2c16-3c78-4164-b2fa-21eef5eff40d","ELEMENT":"dcdf2c16-3c78-4164-b2fa-21eef5eff40d"}}]
1517313740761	geckodriver::marionette	TRACE	<- [1,8,null,{"value":{"element-6066-11e4-a52e-4f735466cecf":"dcdf2c16-3c78-4164-b2fa-21eef5eff40d","ELEMENT":"dcdf2c16-3c78-4164-b2fa-21eef5eff40d"}}]
1517313740762	webdriver::server	DEBUG	<- 200 OK {"value":{"element-6066-11e4-a52e-4f735466cecf":"dcdf2c16-3c78-4164-b2fa-21eef5eff40d"}}
1517313740769	webdriver::server	DEBUG	-> POST /session/d96b0092-df51-4111-913c-771c8623e6d6/element/dcdf2c16-3c78-4164-b2fa-21eef5eff40d/value {"id":"dcdf2c16-3c78-4164-b2fa-21eef5eff40d","text":"Hello","value":["H","e","l","l","o"]}
1517313740770	geckodriver::marionette	TRACE	-> 116:[0,9,"sendKeysToElement",{"id":"dcdf2c16-3c78-4164-b2fa-21eef5eff40d","text":"Hello","value":["H","e","l","l","o"]}]
1517313740772	Marionette	TRACE	0 -> [0,9,"sendKeysToElement",{"id":"dcdf2c16-3c78-4164-b2fa-21eef5eff40d","text":"Hello","value":["H","e","l","l","o"]}]
1517313740788	Marionette	TRACE	0 <- [1,9,null,{}]
1517313740789	geckodriver::marionette	TRACE	<- [1,9,null,{}]
1517313740789	webdriver::server	DEBUG	<- 200 OK {"value": {}}
1517313740798	webdriver::server	DEBUG	-> DELETE /session/d96b0092-df51-4111-913c-771c8623e6d6/window 
1517313740799	geckodriver::marionette	TRACE	-> 17:[0,10,"close",{}]
1517313740803	Marionette	TRACE	0 -> [0,10,"close",{}]
[DEBUG SHUTDOWN] Shutdown: decoder=00000239AC5AE800 state machine=00000239A50B8000
[DEBUG SHUTDOWN] Enter: state machine=000002399F7F1580 reader=00000239A50B9040
[DEBUG SHUTDOWN] Shutdown: reader=00000239A69DB000 shutdown demuxer=00000239AC59EFE0
[DEBUG SHUTDOWN] Shutdown: pool=00000239A69D5DC0 count=1
[DEBUG SHUTDOWN] operator (): pool=00000239A69D5DC0 shutdown=true count=0
[DEBUG SHUTDOWN] TearDownDecoders: reader=00000239A69DB000 shut down audio task queue
[DEBUG SHUTDOWN] TearDownDecoders: reader=00000239A69DB000 shut down video task queue
[DEBUG SHUTDOWN] FinishShutdown: state machine=00000239A50B8000
[DEBUG SHUTDOWN] Unregister: decoder=00000239AC5AE800, count=0


[Test.zip](https://github.com/mozilla/geckodriver/files/1677554/Test.zip)

**Comment by soumyababy81** [comment]
@whimboo Any findings on above?

**Comment by soumyababy81** [comment]
@whimboo Happy to know that the bug is filed for first scenario. Please let us know if any findings on the second scenario - scenario explained with sample tests attached with stack trace?

**Comment by soumyababy81** [comment]
@whimboo any updates on this?

**Comment by whimboo** [comment]
Sorry, I missed to add that this is https://bugzilla.mozilla.org/show_bug.cgi?id=1434872 now.

**Comment by soumyababy81** [comment]
Hi,
 Any updates on this?

**Comment by whimboo** [comment]
Closing because it should be fixed as pointed out in my last comment.
